src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -199,28 +199,33 @@ private void UpdateSize()
                 if (extraField.Size < sizeof(long))
                     return true;
 
-                long value64 = reader.ReadInt64();
+                // Advancing the stream (by reading from it) is possible only when:
+                // 1. There is an explicit ask to do that (valid files, corresponding boolean flag(s) set to true, #77159).
+                // 2. When the size indicates that all the information is available ("slightly invalid files", #49580).
+                bool readAllFields = extraField.Size >= sizeof(long) + sizeof(long) + sizeof(long) + sizeof(int);
+
+                long value64 = readUncompressedSize || readAllFields ? reader.ReadInt64() : -1;
```

Comment:
Could this fail if only 2 or 3 fields are available for reading but the bool params are false?

i.e: what if extraField.size = 16 and readUncompressedSize = false and readCompressedSize = false?

Reply:
I don't think that this is possible. @danmoseley has added following comment:

https://github.com/dotnet/runtime/blob/50b65de0d6deaa45fd48e6991fedb1a029e9af28/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs#L190-L197

My understanding is that we have two options:

1. Correct archives: the fields are provided when the directory record fields are set (the right booleans are set to true, the size is their total size)
2. "Slightly incorrect archives": the boolean fields are not set, but the provided size suggests that they are all present. Tests added in https://github.com/dotnet/runtime/pull/68106 use such approach.